cpp/include/raft/linalg/detail/lanczos.hpp

```diff
@@ -0,0 +1,1400 @@
+// =========================================================
+
+/**
+ *  @brief  Perform Lanczos iteration
+ *    Lanczos iteration is performed on a shifted matrix A+shift*I.
+ *  @tparam index_type_t the type of data used for indexing.
+ *  @tparam value_type_t the type of data used for weights, distances.
+ *  @param handle the raft handle.
+ *  @param A Matrix.
+ *  @param iter Pointer to current Lanczos iteration. On exit, the
+ *    variable is set equal to the final Lanczos iteration.
+ *  @param maxIter Maximum Lanczos iteration. This function will
+ *    perform a maximum of maxIter-*iter iterations.
+ *  @param shift Matrix shift.
+ *  @param tol Convergence tolerance. Lanczos iteration will
+ *    terminate when the residual norm (i.e. entry in beta_host) is
+ *    less than tol.
+ *  @param reorthogonalize Whether to reorthogonalize Lanczos
+ *    vectors.
+ *  @param alpha_host (Output, host memory, maxIter entries)
+ *    Diagonal entries of Lanczos system.
+ *  @param beta_host (Output, host memory, maxIter entries)
+ *    Off-diagonal entries of Lanczos system.
+ *  @param lanczosVecs_dev (Input/output, device memory,
+ *    n*(maxIter+1) entries) Lanczos vectors. Vectors are stored as
+ *    columns of a column-major matrix with dimensions
+ *    n x (maxIter+1).
+ *  @param work_dev (Output, device memory, maxIter entries)
+ *    Workspace. Not needed if full reorthogonalization is disabled.
+ *  @return Zero if successful. Otherwise non-zero.
+ */
+template <typename index_type_t, typename value_type_t>
+int performLanczosIteration(handle_t const& handle,
+                            sparse_matrix_t<index_type_t, value_type_t> const* A,
+                            index_type_t* iter,
+                            index_type_t maxIter,
+                            value_type_t shift,
+                            value_type_t tol,
+                            bool reorthogonalize,
+                            value_type_t* __restrict__ alpha_host,
+                            value_type_t* __restrict__ beta_host,
+                            value_type_t* __restrict__ lanczosVecs_dev,
+                            value_type_t* __restrict__ work_dev)
+{
+  // -------------------------------------------------------
+  // Variable declaration
+  // -------------------------------------------------------
+
+  // Useful variables
+  constexpr value_type_t one    = 1;
+  constexpr value_type_t negOne = -1;
+  constexpr value_type_t zero   = 0;
+  value_type_t alpha;
+
+  auto cublas_h = handle.get_cublas_handle();
+  auto stream   = handle.get_stream();
+
+  RAFT_EXPECTS(A != nullptr, "Null matrix pointer.");
+
+  index_type_t n = A->nrows_;
+
+  // -------------------------------------------------------
+  // Compute second Lanczos vector
+  // -------------------------------------------------------
+  if (*iter <= 0) {
+    *iter = 1;
+
+    // Apply matrix
+    if (shift != 0)
+      CUDA_TRY(cudaMemcpyAsync(lanczosVecs_dev + n,
```

Comment:
We need to be mindful that we update these as well while we're moving this stuff over. `RAFT_CUDA_TRY`, `RAFT_CUBLAS_TRY`, etc...